### ISSUE TITLE
Retention are not working

### ISSUE THREAD

**Issue by Bastilla123** [pgbackrest/pgbackrest#2103]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.45

2. PostgreSQL version: 13.9

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each:

Debian 11

4. Did you install pgBackRest from source or from a package? 

I install from source

5. Please attach the following as applicable:
    - `pgbackrest.conf` file(s)
    - `postgresql.conf` settings applicable to pgBackRest (`archive_command`, `archive_mode`, `listen_addresses`, `max_wal_senders`, `wal_level`, `port`)
    - errors in the postgresql log file before or during the time you experienced the issue
    - log file in `/var/log/pgbackrest` for the commands run (e.g. `/var/log/pgbackrest/mystanza_backup.log`)

6. Describe the issue:

I setup in pgbackrest.conf:

repo1-retention-full-type=time
repo1-retention-full=1

expected behaviour is that Fullbackup and WAL Files which is older then 1 day are deleted. But after 3 days now i see 3 Fullbackups and also WAL Archives over 3 days.

[Config.zip](https://github.com/pgbackrest/pgbackrest/files/11891154/Config.zip)

**Comment by dwsteele** [comment]
Please provide output of `pgbackrest info`.

**Comment by Bastilla123** [comment]
I start pgbackrest manual here output:

:/var/log/pgbackrest$ pgbackrest --stanza=main --type=full --log-level-console=info backup
2023-06-28 12:57:55.082 P00   INFO: backup command begin 2.45: --exec-id=165288-c174b458 --log-level-console=info --pg1-path=/var/lib/postgresql/13/main --pg1-port=5432 --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=main --start-fast --type=full
2023-06-28 12:57:55.800 P00   INFO: execute non-exclusive backup start: backup begins after the requested immediate checkpoint completes
2023-06-28 12:57:56.302 P00   INFO: backup start archive = 000000030000004A00000035, lsn = 4A/35000028
2023-06-28 12:57:56.302 P00   INFO: check archive for prior segment 000000030000004A00000034
WARN: resumable backup 20230628-125728F of same type exists -- invalid files will be removed then the backup will resume
2023-06-28 12:58:47.681 P00   INFO: execute non-exclusive backup stop and wait for all WAL segments to archive
2023-06-28 12:58:47.884 P00   INFO: backup stop archive = 000000030000004A00000036, lsn = 4A/36000088
2023-06-28 12:58:47.903 P00   INFO: check archive for segment(s) 000000030000004A00000035:000000030000004A00000036
2023-06-28 12:58:47.962 P00   INFO: new backup label = 20230628-125728F
2023-06-28 12:58:48.469 P00   INFO: full backup size = 98.3MB, file total = 5378
2023-06-28 12:58:48.469 P00   INFO: backup command end: completed successfully (53387ms)
2023-06-28 12:58:48.469 P00   INFO: expire command begin 2.45: --exec-id=165288-c174b458 --log-level-console=info --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=main
2023-06-28 12:58:48.477 P00   INFO: repo1: expire time-based backup 20230625-220002F
2023-06-28 12:58:48.480 P00   INFO: repo1: expire time-based backup 20230626-220001F
2023-06-28 12:58:48.502 P00   INFO: repo1: remove expired backup 20230626-220001F
2023-06-28 12:58:53.225 P00   INFO: repo1: remove expired backup 20230625-220002F
2023-06-28 12:58:58.849 P00   INFO: repo1: 13-1 remove archive, start = 0000000300000045, stop = 0000000300000046000000DF
2023-06-28 12:58:58.850 P00   INFO: expire command end: completed successfully (10381ms)

Strange now it is working and delete backup from last 2 days which is right but when i see in main_expire.log i find:

-------------------PROCESS START-------------------
2023-06-27 08:42:30.251 P00   INFO: expire command begin 2.45: --exec-id=137553-5025273a --log-level-console=error --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=m>
2023-06-27 08:42:30.268 P00   INFO: repo1: time-based archive retention not met - archive logs will not be expired
2023-06-27 08:42:30.268 P00   INFO: expire command end: completed successfully (17ms)

-------------------PROCESS START-------------------
2023-06-27 22:00:56.241 P00   INFO: expire command begin 2.45: --exec-id=150432-64da3c7c --log-level-console=error --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=m>
2023-06-27 22:00:56.258 P00   INFO: repo1: time-based archive retention not met - archive logs will not be expired
2023-06-27 22:00:56.258 P00   INFO: expire command end: completed successfully (17ms)

-------------------PROCESS START-------------------
2023-06-28 12:58:48.469 P00   INFO: expire command begin 2.45: --exec-id=165288-c174b458 --log-level-console=info --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=ma>
2023-06-28 12:58:48.477 P00   INFO: repo1: expire time-based backup 20230625-220002F
2023-06-28 12:58:48.480 P00   INFO: repo1: expire time-based backup 20230626-220001F
2023-06-28 12:58:48.502 P00   INFO: repo1: remove expired backup 20230626-220001F
2023-06-28 12:58:53.225 P00   INFO: repo1: remove expired backup 20230625-220002F
2023-06-28 12:58:58.849 P00   INFO: repo1: 13-1 remove archive, start = 0000000300000045, stop = 0000000300000046000000DF
2023-06-28 12:58:58.850 P00   INFO: expire command end: completed successfully (10381ms)

As you see retention time is on both days 1 day retention so yesterday have retention met 20230625-220002F but doesn't work. I observe this in next days and i hope i can reproduce this bug and found hopeful problem

**Comment by dwsteele** [comment]
Looks to be working fine to me. Older backups are not expired until there is a new backup that covers the entire retention period, i.e. is will allow you to do PITR for a full day. Multiple backups might be required to make this happen.

For example, if you have a backup 27 hours old and one 3 hours old, the 27 hour old backup cannot be expired because the 3 hour old backup will not cover a full day of retention.

Also, --repo1-retention-full=1 is an incredibly small setting for time-based retention so it makes the results look a bit odd.

**Comment by bioflash** [comment]
Hi, we got similar question about expiring time-based backup. We are on 2.32. We found that after expiring a backup, the archive still left in the disk

Expire the backup
`sudo -u postgres pgbackrest --set=20230606-191834F --stanza=my-stanza expire --log-level-stderr=erro `

After that, check pgbackrest info

```
stanza: my-stanza
    status: ok
    cipher: none

    db (current)
        wal archive min/max (12): 000000010000000000000003/0000000200000002000000D6

        full backup: 20230628-021046F
            timestamp start/stop: 2023-06-28 02:10:46 / 2023-06-28 02:13:01
            wal start/stop: 0000000200000002000000C7 / 0000000200000002000000C7
            database size: 12.7GB, database backup size: 12.7GB
            repo1: backup set size: 1.5GB, backup size: 1.5GB

        full backup: 20230629-175246F
            timestamp start/stop: 2023-06-29 17:52:46 / 2023-06-29 18:00:14
            wal start/stop: 0000000200000002000000D6 / 0000000200000002000000D6
            database size: 12.8GB, database backup size: 12.8GB
            repo1: backup set size: 1.5GB, backup size: 1.5GB
```

The oldest full backup now is taken on 6/28. However if we list the archive files

```
ls -ltr /crypt/pgbackrest_repo/archive/my-stanza/12-1/
total 88
drwxr-x--- 2 postgres postgres 24576 Jun  6 19:39 0000000100000000
drwxr-x--- 2 postgres postgres 24576 Jun  6 19:43 0000000100000001
drwxr-x--- 2 postgres postgres 20480 Jun  6 22:17 0000000100000002
-rw-r----- 1 postgres postgres    68 Jun  6 22:24 00000002.history
drwxr-x--- 2 postgres postgres  4096 Jun 29 18:00 0000000200000002
```

The archive created on 6/6 still there and the expire log says it can not find archive to delete

```
2023-06-29 18:32:45.068 P00   INFO: expire command begin 2.32: --exec-id=7485-6092ff8a --log-level-stderr=error --repo1-path=/crypt/pgbackrest_repo --repo1-retention-full=7 --repo1-retention-full-type=time --set=20230606-191834F --stanza=my-stanza
2023-06-29 18:32:45.072 P00   INFO: expire adhoc backup set repo1: 20230606-191834F, 20230606-191834F_20230606-231136I
2023-06-29 18:32:45.087 P00   INFO: remove expired backup repo1: 20230606-191834F_20230606-231136I
2023-06-29 18:32:45.328 P00   INFO: remove expired backup repo1: 20230606-191834F
2023-06-29 18:32:45.662 P00   INFO: time-based archive retention not met for repo1 - archive logs will not be expired
2023-06-29 18:32:45.662 P00   INFO: expire command end: completed successfully (595ms)

-------------------PROCESS START-------------------
2023-06-29 18:51:48.052 P00   INFO: expire command begin 2.32: --exec-id=9325-ea26cd74 --repo1-path=/crypt/pgbackrest_repo --repo1-retention-full=7 --repo1-retention-full-type=time --stanza=my-stanza
2023-06-29 18:51:48.071 P00   INFO: time-based archive retention not met for repo1 - archive logs will not be expired
2023-06-29 18:51:48.071 P00   INFO: expire command end: completed successfully (20ms)
```

This is our pgbackrest config
```
[global]
repo1-path=/crypt/pgbackrest_repo
repo1-retention-full=7
repo1-retention-full-type=time
compress-type=zst

[my-stanza]
pg1-path=/var/lib/pgsql/12/data
```

Do we miss anything?

**Comment by Bastilla123** [comment]
Now it works fine for me... Strange i don't change anything... but now i can't reproduce

**Comment by dwsteele** [comment]
@bioflash Archive is not expired until retention has been met. This is so there is enough archive to feed replicas. Looks like you did not do any backups for a while or you changed retention. One you have 7 days worth of backups the old WAL should expire.

**Comment by Bastilla123** [comment]
In last days all works fine. Now suddenly Retention not work. Here are log:

-------------------PROCESS START-------------------
2023-07-01 22:00:57.680 P00   INFO: expire command begin 2.45: --exec-id=331302-14a754ca --log-level-console=info --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=ma>
2023-07-01 22:00:57.690 P00   INFO: repo1: expire time-based backup 20230629-220001F
2023-07-01 22:00:57.933 P00   INFO: repo1: remove expired backup 20230629-220001F
2023-07-01 22:01:20.736 P00   INFO: repo1: 13-1 remove archive, start = 000000030000004F, stop = 0000000300000050000000EE
2023-07-01 22:01:20.737 P00   INFO: expire command end: completed successfully (23057ms)

-------------------PROCESS START-------------------
2023-07-02 22:00:55.750 P00   INFO: expire command begin 2.45: --exec-id=412725-76663ebf --log-level-console=info --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=ma>
2023-07-02 22:00:55.766 P00   INFO: repo1: expire time-based backup 20230630-220001F
2023-07-02 22:00:55.786 P00   INFO: repo1: remove expired backup 20230630-220001F
2023-07-02 22:01:01.006 P00   INFO: repo1: 13-1 remove archive, start = 0000000300000052, stop = 0000000300000053000000C0
2023-07-02 22:01:01.007 P00   INFO: expire command end: completed successfully (5257ms)

-------------------PROCESS START-------------------
2023-07-03 22:00:54.996 P00   INFO: expire command begin 2.45: --exec-id=462789-29ce7ee0 --log-level-console=info --repo1-path=/media/backup --repo1-retention-full=1 --repo1-retention-full-type=time --stanza=ma>
2023-07-03 22:00:55.024 P00   INFO: repo1: time-based archive retention not met - archive logs will not be expired
2023-07-03 22:00:55.024 P00   INFO: expire command end: completed successfully (28ms)

**Comment by dwsteele** [comment]
Again, we need output from the `pgbackrest info` command and current logs to be able to look at this.

**Comment by Bastilla123** [comment]
Here are ls of dirs:

root@icinga:/media/backup/archive/main/13-1# ls -lh
insgesamt 0
drwxr-xr-x 2 postgres postgres 0  4. Jul 10:41 0000000300000056
drwxr-xr-x 2 postgres postgres 0  3. Jul 10:06 0000000300000057
drwxr-xr-x 2 postgres postgres 0  3. Jul 18:38 0000000300000058
drwxr-xr-x 2 postgres postgres 0  4. Jul 03:06 0000000300000059
drwxr-xr-x 2 postgres postgres 0  4. Jul 11:34 000000030000005A
drwxr-xr-x 2 postgres postgres 0  4. Jul 20:06 000000030000005B
drwxr-xr-x 2 postgres postgres 0  5. Jul 04:34 000000030000005C
drwxr-xr-x 2 postgres postgres 0  5. Jul 08:56 000000030000005D

root@icinga:/media/backup/backup/main# ls -lh
insgesamt 12K
drwxr-xr-x 2 postgres postgres    0  2. Jul 22:00 20230702-220002F
drwxr-xr-x 2 postgres postgres    0  3. Jul 22:00 20230703-220002F
drwxr-xr-x 2 postgres postgres    0  4. Jul 10:41 20230704-104048F
drwxr-xr-x 2 postgres postgres    0  4. Jul 22:00 20230704-220001F
drwxr-xr-x 2 postgres postgres    0 23. Jun 22:00 backup.history
-rwxr-xr-x 1 postgres postgres 3,0K  4. Jul 22:00 backup.info
-rwxr-xr-x 1 postgres postgres 3,0K  4. Jul 22:00 backup.info.copy
lrwxrwxrwx 1 postgres postgres   16  4. Jul 22:00 latest -> 20230704-220001F

bgpackrest info:

root@icinga:/media/backup/backup/main# pgbackrest info
stanza: main
    status: ok
    cipher: none

    db (current)
        wal archive min/max (13): 000000030000005600000093/000000030000005D00000084

        full backup: 20230702-220002F
            timestamp start/stop: 2023-07-02 22:00:02 / 2023-07-02 22:00:55
            wal start/stop: 000000030000005600000093 / 000000030000005600000094
            database size: 98.5MB, database backup size: 98.5MB
            repo1: backup set size: 10.6MB, backup size: 10.6MB

        full backup: 20230703-220002F
            timestamp start/stop: 2023-07-03 22:00:02 / 2023-07-03 22:00:54
            wal start/stop: 000000030000005900000066 / 000000030000005900000066
            database size: 98.5MB, database backup size: 98.5MB
            repo1: backup set size: 10.6MB, backup size: 10.6MB

        full backup: 20230704-104048F
            timestamp start/stop: 2023-07-04 10:40:48 / 2023-07-04 10:41:41
            wal start/stop: 000000030000005A000000E4 / 000000030000005A000000E4
            database size: 98.6MB, database backup size: 98.6MB
            repo1: backup set size: 10.6MB, backup size: 10.6MB

        full backup: 20230704-220001F
            timestamp start/stop: 2023-07-04 22:00:01 / 2023-07-04 22:00:53
            wal start/stop: 000000030000005C00000039 / 000000030000005C0000003A
            database size: 98.6MB, database backup size: 98.6MB
            repo1: backup set size: 10.6MB, backup size: 10.6MB


Log Files i have added

[a.zip](https://github.com/pgbackrest/pgbackrest/files/11954800/a.zip)

**Comment by dwsteele** [comment]
This looks correct to me. You have one day of retention configured but 20230703-220002F misses that by one second so everything from 20230702-220002F needs to be retained.

It just looks a bit weird because you have the retention so ridiculously low. If it were set to a reasonable value like 14 (which I consider to be bare minimum) then this effect would not be so noticeable.

**Comment by dwsteele** [comment]
However, it was not ideal that you were getting this message:
```
time-based archive retention not met - archive logs will not be expired
```
Harmless, but definitely misleading. We have fixed that in #2109.

**Comment by Bastilla123** [comment]
I think now we can close this issue. I am satisfied
